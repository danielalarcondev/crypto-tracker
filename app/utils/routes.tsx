export enum Routes {
	"ASSETS" = "/assets",
	"DEFAULT" = "/tracker",
	"TRACKER" = "/tracker"
}

export const getDefaultRoute = () => Routes.DEFAULT;