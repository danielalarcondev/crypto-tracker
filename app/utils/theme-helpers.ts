
interface ThemeHelper {
	getIsDarkTheme: () => boolean;
    isDarkThemeConfigured: () => boolean;
    // eslint-disable-next-line no-unused-vars
    setIsDarkTheme: (shouldSetToDark: boolean) => void;
}

const IS_DARK_THEME_KEY = 'IS_DARK_THEME';

export const isDarkThemeConfigured = () => localStorage?.getItem(IS_DARK_THEME_KEY) !== null;
export const getIsDarkTheme = () => Boolean(localStorage?.getItem(IS_DARK_THEME_KEY));

export const setIsDarkTheme = (shouldSetToDark: boolean) => {
    localStorage?.setItem(IS_DARK_THEME_KEY, shouldSetToDark.toString());
};

const themeLocalStorage: ThemeHelper = {
    getIsDarkTheme,
    isDarkThemeConfigured,
    setIsDarkTheme
};

export default themeLocalStorage;
 
