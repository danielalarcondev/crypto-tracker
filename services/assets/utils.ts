import { GetAssetsResponse } from '@common-types/assets';

export const DEFAULT_ASSETS_LIMIT = 100;
export const DEFAULT_ASSETS_OFFSET = 0;

export const DEFAULT_ERROR_CODE = 500;
export const DEFAULT_ERROR_MESSAGE = 'Something went wrong';
export const URL_ERROR_CODE = 404;
export const URL_ERROR_MESSAGE = 'Server could not be reached. Try again later';

export const defaultErrorResponse: GetAssetsResponse = {
    data: [],
    timestamp: Date.now(),
    error: { 
        errorMessage: DEFAULT_ERROR_MESSAGE, 
        httpError: DEFAULT_ERROR_CODE
    }
};

export const urlErrorResponse: GetAssetsResponse = {
    data: [],
    timestamp: Date.now(),
    error: { 
        errorMessage: URL_ERROR_MESSAGE,
        httpError: URL_ERROR_CODE
    }
};
