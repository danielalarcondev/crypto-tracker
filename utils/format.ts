export const formatLargeNumber = (number: string) => {

    const formatter: Intl.NumberFormat = new Intl.NumberFormat('en-US', {
        notation: 'compact',
        compactDisplay: 'short',
        minimumFractionDigits: 2,
        maximumFractionDigits: 2
    });

    return formatter.format(Number(number));
};

export const formatPercentageNumber = (percentage: string) => {

    const TWO_DECIMALS = 2;
    const number = Number(percentage);

    if (isNaN(number)) {
        return `${percentage}%`;
    }

    return `${number.toFixed(TWO_DECIMALS)}%`;
};
