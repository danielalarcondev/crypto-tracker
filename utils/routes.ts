export enum Routes {
	"ASSETS" = "/assets",
	"DEFAULT" = "/tracker",
	"TRACKER" = "/tracker"
}
