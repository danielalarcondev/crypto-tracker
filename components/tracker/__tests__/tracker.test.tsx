import * as React from 'react';
import '@testing-library/jest-dom/jest-globals';

import { render, screen, waitFor } from '@testing-library/react';
import { expect, it, describe, beforeEach } from '@jest/globals';
import Tracker from '@components/tracker/tracker';
import { mockedAssets } from '@tests/utils';

describe('TrackerPage: ', () => {

    beforeEach(() => (
        global.ResizeObserver = jest.fn().mockImplementation(() => ({
            observe: jest.fn(),
            unobserve: jest.fn(),
            disconnect: jest.fn(),
        }))
    ));

    it('should render correctly', async () => {

        const mockedResponse = {
            data: mockedAssets,
            timestamp: Date.now(),
        };

        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: true,
                json: () => Promise.resolve({
                    ...mockedResponse
                }),
            })
        );

        render(<Tracker />);

        let skeletonloading = screen.queryByTestId('tracker-table-skeleton');
        expect(skeletonloading).toBeInTheDocument();

        await waitFor(() => {
            const pagination = screen.getByTestId('tracker-pagination');
            const table = screen.getByTestId('tracker-table');
            skeletonloading = screen.queryByTestId('tracker-table-skeleton');
	
            expect(pagination).toBeInTheDocument();
            expect(table).toBeInTheDocument();
            expect(skeletonloading).toBeFalsy();
        });
        
    });

    it('should render error page if service returns no assets', async () => {

        const statusText = 'Custom error';
        const status = 508;
		
        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: false,
                statusText,
                status
            })
        );
		
        const { container } = render(<Tracker />);

        let loading = screen.queryByTestId('tracker-table-skeleton');
        let noResultsError = screen.queryByTestId('assents-no-results');

        expect(loading).toBeInTheDocument();
        expect(noResultsError).toBeFalsy();

        await waitFor(() => {
            noResultsError = screen.queryByTestId('assents-no-results');
            loading = screen.queryByTestId('tracker-table-skeleton');
	
            expect(noResultsError).toBeInTheDocument();
            expect(loading).toBeFalsy();
            expect(container).toMatchSnapshot();
        });
        
    });
});
