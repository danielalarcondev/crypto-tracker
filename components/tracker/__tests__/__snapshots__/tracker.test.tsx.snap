// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerPage:  should render error page if service returns no assets 1`] = `
<div>
  <div
    class="flex flex-col gap-6 w-full"
    data-testid="tracker"
  >
    <div
      class="rounded-lg border border-neutral-300 bg-neutral-100 dark:bg-neutral-600"
    />
    <div
      class="w-[26rem] h-32 bg-red-300 rounded-md p-6"
      data-testid="assents-no-results"
    >
      <h1
        class="text-md mb-4"
        data-testid="assents-no-results-message"
      >
        No results. Would you like to try again?
      </h1>
      <button
        class="group relative flex items-stretch justify-center p-0.5 text-center font-medium transition-[color,background-color,border-color,text-decoration-color,fill,stroke,box-shadow] focus:z-10 focus:outline-none bg-blue-500/80 hover:bg-blue-500 border border-white text-white rounded-lg w-full"
        data-testid="assets-no-results-retry-button"
        type="button"
      >
        <span
          class="flex items-stretch transition-all duration-200 rounded-md px-4 py-2 text-sm"
        >
          Retry
          <svg
            class="w-4 h-4 ml-2 flex self-center"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0 3.181 3.183a8.25 8.25 0 0 0 13.803-3.7M4.031 9.865a8.25 8.25 0 0 1 13.803-3.7l3.181 3.182m0-4.991v4.99"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="Toastify"
    >
      <div
        class="Toastify__toast-container Toastify__toast-container--top-center"
      >
        <div
          class="Toastify__toast Toastify__toast-theme--light Toastify__toast--error Toastify--animate Toastify__bounce-enter--top-center"
          data-in="true"
          id="[object Object]"
        >
          <div
            class="Toastify__toast-body"
            role="alert"
          >
            <div
              class="Toastify__toast-icon Toastify--animate-icon Toastify__zoom-enter"
            >
              <svg
                fill="var(--toastify-icon-color-error)"
                height="100%"
                viewBox="0 0 24 24"
                width="100%"
              >
                <path
                  d="M11.983 0a12.206 12.206 0 00-8.51 3.653A11.8 11.8 0 000 12.207 11.779 11.779 0 0011.8 24h.214A12.111 12.111 0 0024 11.791 11.766 11.766 0 0011.983 0zM10.5 16.542a1.476 1.476 0 011.449-1.53h.027a1.527 1.527 0 011.523 1.47 1.475 1.475 0 01-1.449 1.53h-.027a1.529 1.529 0 01-1.523-1.47zM11 12.5v-6a1 1 0 012 0v6a1 1 0 11-2 0z"
                />
              </svg>
            </div>
            <div>
              Custom error
            </div>
          </div>
          <button
            aria-label="close"
            class="Toastify__close-button Toastify__close-button--light"
            type="button"
          >
            <svg
              aria-hidden="true"
              viewBox="0 0 14 16"
            >
              <path
                d="M7.71 8.23l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75L1 11.98l3.75-3.75L1 4.48 2.48 3l3.75 3.75L9.98 3l1.48 1.48-3.75 3.75z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <div
            class="Toastify__progress-bar--wrp"
            data-hidden="false"
          >
            <div
              class="Toastify__progress-bar--bg Toastify__progress-bar-theme--light Toastify__progress-bar--error"
            />
            <div
              aria-hidden="false"
              aria-label="notification timer"
              class="Toastify__progress-bar Toastify__progress-bar--animated Toastify__progress-bar-theme--light Toastify__progress-bar--error"
              role="progressbar"
              style="animation-duration: 2000ms; animation-play-state: paused;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
